include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td

```diff
@@ -161,6 +161,57 @@ def InstanceOp : HardwareDeclOp<"instance", [
   }];
 }
 
+def InstanceChoiceOp : HardwareDeclOp<"instance_choice", [
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>
+  ]> {
+  let summary = "Creates an instance of a module based on a option";
+
+  let description = [{
+    The instance choice operation creates an instance choosing the target based
+    on the value of an option if one is specified, instantiating a default
+    target otherwise.
+
+    The port lists of all instance targets must match exactly.
+
+    Examples:
+    ```mlir
+    %0 = firrtl.instance_choice foo @Foo,
+      @FpgaFoo if @Opt::@FPGA
+      (in io: !firrtl.uint)
+    ```
+  }];
+
+  let arguments = (ins FlatSymbolRefArrayAttr:$moduleNames,
+                       SymbolRefArrayAttr:$caseNames,
```

Comment:
Does the `moduleNames` argument include the default name?

This places a somewhat odd constraint that moduleNames needs to be sized one larger than caseNames. Alternatively, the PR could split this into two arrays for the names, `defaultModuleName` and `alternativeModuleNames`. There may be a trait that can then tell ODS that these have to be the same size. Regardless, the validity of the sizes should be checked.

Relatedly, it seems odd to use a different representation for the arguments and for the build with respect to moduleNames.

If in subsequent code that uses this, if keeping the current representation, it would be good to add a helper to get the default module as opposed to a lot of `getModulesNames()[0]`. Keep an eye out for this.

Reply:
Keeping it as a unified array seems fine. I agree that this is closer to the true representation: this can be any of these things. In that sense the default isn't really special.